eclipse/src/saros/editor/internal/ContributionAnnotationManager.java

```diff
@@ -228,88 +253,112 @@ public void refreshAnnotations(IAnnotationModel model) {
+      annotation.getModel().addAnnotation(annotation, position);
+      return;
     }
+
+    // TODO OPTIMIZE it is usually 2 annotations after the queue is full
+    annotation.getModel().addAnnotation(annotation, position);
+
+    for (final HistoryEntry entry : removedEntries)
+      entry.annotation.getModel().removeAnnotation(entry.annotation);
   }
 
   /**
-   * Removes an annotation from the user's history and the annotation model.
-   *
-   * @param annotation
+   * Removes entries from the history based on the history current insert stamp. This method <b>DOES
+   * NOT</b> alter the annotation model the removed annotations in the history belong to!
    */
-  private void removeFromHistory(ContributionAnnotation annotation) {
-    getHistory(annotation.getSource()).remove(annotation);
-    annotation.getModel().removeAnnotation(annotation);
+  private List<HistoryEntry> removeHistoryEntries(final History history) {
+
+    final List<HistoryEntry> removedEntries = new ArrayList<>();
+
+    final int insertStampToRemove = history.currentInsertStamp;
+
+    /* the logic assumes that the entry order does not change during lifetime
+     * i.e if we have a history of size 4 the list must look like this regarding the insert stamps
+     * 0 0 0 1 2 3 3 3 0 1 1 1 2 2 2 2 3 0 1 1 1 1 1 2 3 ... and so on
+     *
+     */
+
+    final Iterator<HistoryEntry> it = history.entries.iterator();
+
+    if (!it.hasNext()) return removedEntries;
+
+    while (it.hasNext()) {
+      final HistoryEntry entry = it.next();
+
+      if (entry.insertStamp != insertStampToRemove) break;
+
+      removedEntries.add(entry);
+      it.remove();
+    }
+
+    return removedEntries;
   }
 
   /**
-   * Replaces an existing annotation in the current history with a new annotation.
-   *
-   * @param oldAnnotation
-   * @param newAnnotation
+   * Replaces an existing annotation in the current history with a new annotation. This method
+   * <b>DOES NOT</b> alter the annotation model!
    */
   private void replaceInHistory(
-      ContributionAnnotation oldAnnotation, ContributionAnnotation newAnnotation) {
+      final ContributionAnnotation oldAnnotation, final ContributionAnnotation newAnnotation) {
+
     assert oldAnnotation.getSource().equals(newAnnotation.getSource());
 
-    LinkedList<ContributionAnnotation> list = sourceToHistory.get(oldAnnotation.getSource());
+    final History history = sourceToHistory.get(oldAnnotation.getSource());
 
-    if (list == null) {
+    if (history == null) {
       log.warn("a annotation history for user " + oldAnnotation.getSource() + " does not exists");
 
       return;
     }
 
-    for (ListIterator<ContributionAnnotation> it = list.listIterator(); it.hasNext(); ) {
-      ContributionAnnotation annotation = it.next();
-      if (annotation.equals(oldAnnotation)) {
-        it.set(newAnnotation);
+    for (final ListIterator<HistoryEntry> it = history.entries.listIterator(); it.hasNext(); ) {
```

Comment:
Okay, am I missing something at this point? What is the matter with all those manual iterator loops?

Reply:
Well in this case I could you for:each.

I guess my intention was to replace the HistoryEntry but it seems I just replaced the annotation in the entry as this works as well.